File: langs/csharp/src/VMx/Collections/TokenPagedComposition.cs
using System.Collections.Specialized;
using System.ComponentModel;
using System.Reactive;
using System.Reactive.Subjects;
using VMx.Commands;
using VMx.Components;

namespace VMx.Collections;

/// <summary>A single page returned by <see cref="TokenPagedComposition{TVM,TToken}"/>.</summary>
public sealed record TokenPage<TVM, TToken>(IReadOnlyList<TVM> Items, TToken? NextToken);

/// <summary>
/// Accumulates pages fetched by an opaque forward-only token.
/// </summary>
public sealed class TokenPagedComposition<TVM, TToken> :
    INotifyCollectionChanged,
    INotifyPropertyChanged,
    IDisposable
{
    private readonly Func<TToken?, Task<TokenPage<TVM, TToken>>> _fetchNext;
    private readonly bool _autoConstructOnAdd;
    private readonly Func<IReadOnlyList<TVM>, IReadOnlyList<TVM>, bool> _pagesEqual;
    private readonly object _gate = new();
    private readonly List<TVM> _items = [];
    // Shared re-signal for BOTH commands: a change driven by one command (e.g.
    // RefreshCommand flipping HasMore) must re-raise the other's CanExecuteChanged,
    // matching Python/TypeScript/Swift (a bound "Load More" re-enables immediately).
    private readonly ISubject<Unit> _commandChanged = Subject.Synchronize(new Subject<Unit>());
    private TToken? _currentToken;
    private bool _loadedOnce;
    private long _operationGeneration;
    private bool _disposed;

    /// <summary>
    /// Creates a token-paged composition.
    /// </summary>
    public TokenPagedComposition(
        Func<TToken?, Task<TokenPage<TVM, TToken>>> fetchNext,
        bool autoConstructOnAdd = false,
        Func<IReadOnlyList<TVM>, IReadOnlyList<TVM>, bool>? pagesEqual = null)
    {
        _fetchNext = fetchNext ?? throw new ArgumentNullException(nameof(fetchNext));
        _autoConstructOnAdd = autoConstructOnAdd;
        _pagesEqual = pagesEqual ?? DefaultPagesEqual;
        LoadMoreCommand = AsyncRelayCommand.Builder()
            .Predicate(CanLoadMore)
            .Task(_ => LoadMoreAsync())
            .Triggers(_commandChanged)
            .Build();
        RefreshCommand = AsyncRelayCommand.Builder()
            .Predicate(IsLive)
            .Task(_ => RefreshAsync())
            .Triggers(_commandChanged)
            .Build();
    }

    /// <summary>Accumulated items loaded so far.</summary>
    public IReadOnlyList<TVM> Items
    {
        get
        {
            lock (_gate)
            {
                return _items.ToArray();
            }
        }
    }

    /// <summary>The token to pass to the next load, or null after the terminal page.</summary>
    public TToken? CurrentToken
    {
        get
        {
            lock (_gate)
            {
                return _currentToken;
            }
        }
    }

    /// <summary>True before the first fetch or while the current token is non-null.</summary>
    public bool HasMore
    {
        get
        {
            lock (_gate)
            {
                return !_loadedOnce || _currentToken is not null;
            }
        }
    }

    /// <summary>Loads the next page and appends it to <see cref="Items"/>.</summary>
    public AsyncRelayCommand LoadMoreCommand { get; }

    /// <summary>Clears/replaces the accumulator by fetching from the initial token.</summary>
    public AsyncRelayCommand RefreshCommand { get; }

    /// <inheritdoc/>
    public event NotifyCollectionChangedEventHandler? CollectionChanged;

    /// <inheritdoc/>
    public event PropertyChangedEventHandler? PropertyChanged;

    private async Task LoadMoreAsync()
    {
        TToken? token;
        long generation;
        lock (_gate)
        {
            if (_disposed) return;
            generation = ++_operationGeneration;
            token = _currentToken;
        }
        var page = await _fetchNext(token).ConfigureAwait(false);
        var additions = page.Items.ToArray();
        lock (_gate)
        {
            if (_disposed || generation != _operationGeneration) return;
        }
        ConstructIfNeeded(additions);
        lock (_gate)
        {
            if (_disposed || generation != _operationGeneration) return;
            _items.AddRange(additions);
            _currentToken = page.NextToken;
            _loadedOnce = true;
        }
        NotifyResetIfLive();
    }

    private async Task RefreshAsync()
    {
        long generation;
        lock (_gate)
        {
            if (_disposed) return;
            generation = ++_operationGeneration;
        }
        var page = await _fetchNext(default).ConfigureAwait(false);
        var fresh = page.Items.ToArray();
        TVM[] head;
        lock (_gate)
        {
            if (_disposed || generation != _operationGeneration) return;
            head = _items.Take(fresh.Length).ToArray();
        }

        var pagesMatch = _pagesEqual(fresh, head);
        if (!pagesMatch)
            ConstructIfNeeded(fresh);

        lock (_gate)
        {
            if (_disposed || generation != _operationGeneration) return;
            if (!pagesMatch)
            {
                _items.Clear();
                _items.AddRange(fresh);
            }
            _currentToken = page.NextToken;
            _loadedOnce = true;
        }

        if (pagesMatch)
            NotifyPropertiesIfLive();
        else
            NotifyResetIfLive();
    }

    private void ConstructIfNeeded(IReadOnlyList<TVM> items)
    {
        if (!_autoConstructOnAdd) return;
        foreach (var item in items)
        {
            if (item is ComponentVMBase vm && !vm.IsConstructed)
                vm.Construct();
        }
    }

    private void NotifyResetIfLive()
    {
        if (!IsLive()) return;
        CollectionChanged?.Invoke(
            this,
            new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
        NotifyPropertiesIfLive();
    }

    private void NotifyPropertiesIfLive()
    {
        foreach (var name in new[] { nameof(Items), nameof(CurrentToken), nameof(HasMore) })
        {
            if (!IsLive()) return;
            OnPropertyChanged(name);
        }
        if (!IsLive()) return;
        // Re-signal both commands' CanExecuteChanged (parity with Python/TS/Swift).
        _commandChanged.OnNext(Unit.Default);
    }

    private bool CanLoadMore()
    {
        lock (_gate)
        {
            return !_disposed && (!_loadedOnce || _currentToken is not null);
        }
    }

    private bool IsLive()
    {
        lock (_gate)
        {
            return !_disposed;
        }
    }

    private void OnPropertyChanged(string name) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));

    private static bool DefaultPagesEqual(IReadOnlyList<TVM> left, IReadOnlyList<TVM> right) =>
        left.SequenceEqual(right, EqualityComparer<TVM>.Default);

    /// <inheritdoc/>
    public void Dispose()
    {
        lock (_gate)
        {
            if (_disposed) return;
            _disposed = true;
        }
        LoadMoreCommand.Dispose();
        RefreshCommand.Dispose();
        _commandChanged.OnCompleted();
    }
}

File: langs/python/src/vmx/collections/token_paged_composition.py
"""TokenPagedComposition — accumulated, forward-only token pagination."""

from __future__ import annotations

from collections.abc import Awaitable, Callable, Sequence
from typing import Generic, TypeVar

import reactivex as rx
from reactivex import operators as ops
from reactivex.subject import Subject

from vmx.collections.collection_changed import CollectionChangedEvent
from vmx.commands.async_relay_command import AsyncRelayCommand
from vmx.components.base import _ComponentVMBase

TVM = TypeVar("TVM")
TToken = TypeVar("TToken")

FetchNext = Callable[[TToken | None], Awaitable[tuple[Sequence[TVM], TToken | None]]]
PagesEqual = Callable[[Sequence[TVM], Sequence[TVM]], bool]


class TokenPagedComposition(Generic[TVM, TToken]):
    """Accumulates pages fetched by an opaque forward-only token."""

    def __init__(
        self,
        fetch_next: FetchNext[TToken, TVM],
        *,
        auto_construct_on_add: bool = False,
        pages_equal: PagesEqual[TVM] | None = None,
    ) -> None:
        self._fetch_next = fetch_next
        self._auto_construct_on_add = auto_construct_on_add
        self._pages_equal = pages_equal or (lambda left, right: list(left) == list(right))
        self._items: list[TVM] = []
        self._current_token: TToken | None = None
        self._loaded_once = False
        self._disposed = False
        self._collection_changed: Subject[CollectionChangedEvent] = Subject()
        self._property_changed: Subject[str] = Subject()
        self._command_changed: Subject[None] = Subject()
        self._load_more_command = (
            AsyncRelayCommand.builder()
            .predicate(lambda: self.has_more and not self._disposed)
            .triggers(self._command_changed)
            .task(self._load_more)
            .build()
        )
        self._refresh_command = (
            AsyncRelayCommand.builder()
            .predicate(lambda: not self._disposed)
            .triggers(self._command_changed)
            .task(self._refresh)
            .build()
        )

    @property
    def items(self) -> list[TVM]:
        return list(self._items)

    @property
    def current_token(self) -> TToken | None:
        return self._current_token

    @property
    def has_more(self) -> bool:
        return not self._loaded_once or self._current_token is not None

    @property
    def load_more_command(self) -> AsyncRelayCommand:
        return self._load_more_command

    @property
    def refresh_command(self) -> AsyncRelayCommand:
        return self._refresh_command

    @property
    def on_collection_changed(self) -> rx.Observable[CollectionChangedEvent]:
        return self._collection_changed.pipe(ops.as_observable())

    @property
    def on_property_changed(self) -> rx.Observable[str]:
        return self._property_changed.pipe(ops.as_observable())

    async def _load_more(self) -> None:
        page, next_token = await self._fetch_next(self._current_token)
        if self._disposed:
            return
        additions = list(page)
        self._items.extend(additions)
        self._construct_if_needed(additions)
        self._current_token = next_token
        self._loaded_once = True
        self._notify_reset()

    async def _refresh(self) -> None:
        page, next_token = await self._fetch_next(None)
        if self._disposed:
            return
        fresh = list(page)
        if self._pages_equal(fresh, self._items[: len(fresh)]):
            self._current_token = next_token
            self._loaded_once = True
            self._notify_properties()
            return
        self._items = fresh
        self._construct_if_needed(fresh)
        self._current_token = next_token
        self._loaded_once = True
        self._notify_reset()

    def _construct_if_needed(self, items: Sequence[TVM]) -> None:
        if not self._auto_construct_on_add:
            return
        for item in items:
            if isinstance(item, _ComponentVMBase) and not item.is_constructed:
                item.construct()

    def _notify_reset(self) -> None:
        self._collection_changed.on_next(CollectionChangedEvent(action="reset"))
        self._notify_properties()

    def _notify_properties(self) -> None:
        for name in ("items", "current_token", "has_more"):
            self._property_changed.on_next(name)
        self._command_changed.on_next(None)

    def dispose(self) -> None:
        if self._disposed:
            return
        self._disposed = True
        self._load_more_command.dispose()
        self._refresh_command.dispose()
        self._collection_changed.on_completed()
        self._collection_changed.dispose()
        self._property_changed.on_completed()
        self._property_changed.dispose()
        self._command_changed.on_completed()
        self._command_changed.dispose()
